### PR TITLE
Strengthen alias elimination

### PR DESCRIPTION
**Pull request by Keno**
Previously, we would only alias eliminate variables that do not occur
differentiated. This was unnecessarily restricted for systems such as
```
x ~ 0
D(x) ~ x - y
```
where alias elimination is now able to determine that `x` and `y` are
aliases.

While this is the main point of the PR, this does improve alias elimination
in some of the tests, causing tearing order to change. Previously we had
a suboptimality in tearing where an equation assigned to an scc in which it
is not able to solve a variable would not be reconsidered in another scc.
Fix that also, to make sure the tests keep working.

### CODE REVIEW

File: src/systems/alias_elimination.jl [hunk 1]
```diff
@@ -287,11 +299,7 @@ function alias_eliminate_graph!(graph, var_to_diff, mm_orig::SparseMatrixCLIL)
     function lss!(ei::Integer)
         vi = pivots[ei]
         may_eliminate = true
-        for v in 𝑠neighbors(graph, mm.nzrows[ei])
-            # the differentiated variable cannot be eliminated
-            may_eliminate &= isnothing(diff_to_var[v]) && isnothing(var_to_diff[v])
-        end
-        locally_structure_simplify!((@view mm[ei, :]), vi, ag, may_eliminate)
+        locally_structure_simplify!((@view mm[ei, :]), vi, ag, var_to_diff)
```
Comment (YingboMa) [inline src/systems/alias_elimination.jl:302]: Could you make sure that variables that are marked irreducible to not be eliminated away? See https://github.com/SciML/ModelingToolkit.jl/blob/083949f0e9c231593a1a4287c7525606f3b27616/src/structural_transformation/utils.jl#L173

**Review by YingboMa — COMMENTED** [review]
Could you test the example
```julia
x ~ 0
D(x) ~ x - y
```
get eliminated as expected? What about
```julia
D(z) ~ 0
```
if we eliminate `D(z)` away, then we won't have `z` as an observed variable as no initial condition would be provided for it.

**Comment by YingboMa** [comment]
BTW, remember to run `using JuliaFormatter, ModelingToolkit; format(joinpath(dirname(pathof(ModelingToolkit)), ".."))` before committing.

**Comment by YingboMa** [comment]
The CI failures https://github.com/SciML/ModelingToolkit.jl/runs/6804925627?check_suite_focus=true#step:6:1540 look legitimate.

File: src/systems/alias_elimination.jl (hunk 1 above)
Comment (Keno) [reply src/systems/alias_elimination.jl:302]: irreducible values are not included in the linear subsystem, so they should not be eliminated.

File: src/structural_transformation/bipartite_tearing/modia_tearing.jl [hunk 2]
```diff
@@ -37,17 +37,57 @@ end
 
 function tear_graph_modia(structure::SystemStructure; varfilter = v -> true,
                           eqfilter = eq -> true)
+    # It would be possible here to simply iterate over all variables and attempt to
```
Comment (YingboMa) [inline src/structural_transformation/bipartite_tearing/modia_tearing.jl:40]: Should we split this to another PR?
